components/raftstore-v2/src/operation/ready/apply_trace.rs

```diff
@@ -544,6 +555,75 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
         }
         apply_trace.maybe_advance_admin_flushed(apply_index);
     }
+
+    pub fn flush_before_close<T>(&mut self, ctx: &StoreContext<EK, ER, T>, tx: SyncSender<()>) {
+        info!(
+            self.logger,
+            "region flush before close begin";
+        );
+        let region_id = self.region_id();
+        let flush_threshold: u64 = (|| {
+            fail_point!("flush_before_cluse_threshold", |t| {
+                t.unwrap().parse::<u64>().unwrap()
+            });
+            50
+        })();
+
+        if let Some(tablet) = self.tablet().cloned() {
+            // flush the oldest cf one by one until we are under the replay count threshold
+            loop {
+                let replay_count = self.storage().estimate_replay_count();
+                if replay_count < flush_threshold {
+                    break;
+                }
+                info!(
+                    self.logger,
+                    "flush-before-close: replay count exceeds threshold, pick the oldest cf to flush";
+                    "count" => replay_count,
+                );
+                tablet.flush_oldest_cf(true, None).unwrap();
+
+                let flush_state = self.flush_state().clone();
+                let mut apply_trace = self.storage_mut().apply_trace_mut();
+                let mut max_flush_index = 0;
+
+                let flushed_indexes = flush_state.as_ref().flushed_index();
+                for i in 0..flushed_indexes.len() {
+                    let flush_index = flushed_indexes[i].load(Ordering::SeqCst);
+                    let cf = offset_to_cf(i);
+                    apply_trace.on_flush(cf, flush_index);
+                    max_flush_index = u64::max(max_flush_index, flush_index);
+                }
+
+                apply_trace.maybe_advance_admin_flushed(max_flush_index);
+                let admin_flush = apply_trace.admin.flushed;
+                apply_trace.persisted_applied = admin_flush;
+
+                if self.storage().estimate_replay_count() < flush_threshold {
+                    let (_, _, tablet_index) = ctx
+                        .tablet_registry
+                        .parse_tablet_name(Path::new(tablet.path()))
+                        .unwrap();
+                    let mut lb = ctx.engine.log_batch(1);
```

Comment:
Why is this necessary? Isn't flushed index written in EventListener? @SpadeA-Tang

Reply:
Only data cf will be flushed in EventListener.

Reply:
That is a design flaw. We should try to advance admin flushed first thing after restart, then use the refreshed index to read raft logs.

Reply:
Also this PR didn't remove the old flush after start logic.

Reply:
> Also this PR didn't remove the old flush-after-start logic.

Only paused peer will trigger flush and I think it's fine.

Reply:
What do you mean? Every region will be flushed once the replay is completed. The code is at `Peer::try_compelete_recovery`.

Reply:
Fixing them here: https://github.com/tikv/tikv/pull/15023

Reply:
> What do you mean? Every region will be flushed once the replay is completed. The code is at `Peer::try_compelete_recovery`.

![image](https://github.com/tikv/tikv/assets/71589810/4ae23160-01b4-4d1c-b8cd-bd1ae5c0a47a)
`pause_for_replay` means the replay count exceeds 128